go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,555 @@
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	invalidIPAddres  = "0.0.0.1"
+	haLeaderLockName = "ovn-kubernetes-master"
+	ovnkubeDbEp      = "ovnkube-db"
+)
+
+//TODO: Need to make this configurable.
+const (
+	ovnNorthDbPort int32 = 6641
+	ovnSouthDbPort int32 = 6642
+)
+
+//TODO: Need to come up with proper default values.
+const (
+	leaseDuration = 6 * time.Second
+	renewDeadline = 3 * time.Second
+	retryPeriod   = 2 * time.Second
+)
+
+// OvnHAMasterController is the object holder for managing the HA master
+// cluster
+type OvnHAMasterController struct {
+	kubeClient      kubernetes.Interface
+	cluster         *OvnClusterController
+	ovnController   *ovn.Controller
+	nodeName        string
+	manageDBServers bool
+	isLeader        bool
+	leaderElector   *leaderelection.LeaderElector
+}
+
+// NewHAMasterController creates a new HA Master controller
+func NewHAMasterController(kubeClient kubernetes.Interface, cluster *OvnClusterController,
+	nodeName string, manageDBServers bool) *OvnHAMasterController {
+	ovnController := ovn.NewOvnController(kubeClient, cluster.watchFactory)
+	return &OvnHAMasterController{
+		kubeClient:      kubeClient,
+		cluster:         cluster,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		leaderElector:   nil,
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backmode mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddres, "6641", "6642")
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterNewLeader := func(nodeName string) {
+		logrus.Infof(nodeName + " is the new leader")
+		wasLeader := hacluster.isLeader
+
+		if hacluster.nodeName == nodeName {
+			logrus.Infof(" I (" + hacluster.nodeName + ") won the election. In active mode")
+			err = hacluster.ConfigureAsActive(nodeName)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				panic(err.Error())
+			}
+			hacluster.isLeader = true
+		} else {
```

Comment:
Is there any chance that ovn-northd could still be running? If, say, the container was restarted and leader was lost? If so, I'd like this to validate that northd isn't running, even if `wasLeader == false`.

Reply:
Since ovnkube is not starting/exec'ing ovn-northd, it should be running right ?
I mean ovn-northd will be started in the daemonset right ?
I can see here that ovn-northd is started this way - https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/ovnkube-master.yaml#L48

From what I understand, ovn-northd will be running on all the API nodes (or ovnkube master nodes). The node which wins election will resume the ovn-northd to be active. The nodes which lost election will pause the ovn-northd.

Reply:
Right - so my concern is that somehow the state transition to stop northd was missed. As a general rule, Kubernetes components should be level-triggered and not rely on individual edges.

Reply:
Ok. I will update the PR soon.

Reply:
@numansiddique was Casey's comment here addressed? If so could you resolve the thread? Thanks!

Reply:
Yes. It's resolved.